src/rgw/rgw_sync_trace.h

```diff
@@ -0,0 +1,183 @@
+#ifndef CEPH_RGW_SYNC_LOG_H
+#define CEPH_RGW_SYNC_LOG_H
+
+#include <atomic>
+
+#include "common/Mutex.h"
+#include "common/RWLock.h"
+#include "common/admin_socket.h"
+
+#include <set>
+#include <ostream>
+#include <string>
+#include <boost/circular_buffer.hpp>
+
+#define SSTR(o) ({      \
+  std::stringstream ss; \
+  ss << o;              \
+  ss.str();             \
+})
+
+#define RGW_SNS_FLAG_ACTIVE   1
+#define RGW_SNS_FLAG_ERROR    2
+
+class RGWRados;
+class RGWSyncTraceManager;
+class RGWSyncTraceNode;
+class RGWSyncTraceNodeContainer;
+class RGWSyncTraceServiceMapThread;
+
+using RGWSyncTraceNodeRef = std::shared_ptr<RGWSyncTraceNode>;
+using RGWSTNCRef = std::shared_ptr<RGWSyncTraceNodeContainer>;
+
+class RGWSyncTraceNode {
+  friend class RGWSyncTraceManager;
+
+  CephContext *cct;
+
+  RGWSyncTraceManager *manager{nullptr};
+  RGWSyncTraceNodeRef parent;
+
+  uint16_t state{0};
+  std::string status;
+
+  Mutex lock{"RGWSyncTraceNode::lock"};
+
+protected:
+  std::string type;
+  std::string id;
+
+  std::string prefix;
+
+  std::string resource_name;
+
+  uint64_t handle;
+
+  boost::circular_buffer<string> history;
+public:
+  RGWSyncTraceNode(CephContext *_cct, RGWSyncTraceManager *_manager, const RGWSyncTraceNodeRef& _parent,
+           const std::string& _type, const std::string& _id);
+
+  void set_resource_name(const string& s) {
+    resource_name = s;
+  }
+
+  const string& get_resource_name() {
+    return resource_name;
+  }
+
+  void set_flag(uint16_t s) {
+    state |= s;
+  }
+  void unset_flag(uint16_t s) {
+    state &= ~s;
+  }
+  bool test_flags(uint16_t f) {
+    return (state & f) == f;
+  }
+  void log(int level, const std::string& s);
+  void finish();
+
+  std::string to_str() {
+    return prefix + " " + status;
+  }
+
+  const string& get_prefix() {
+    return prefix;
+  }
+
+  std::ostream& operator<<(std::ostream& os) { 
+    os << to_str();
+    return os;            
+  }
+
+  boost::circular_buffer<string>& get_history() {
+    return history;
+  }
+
+  bool match(const string& search_term, bool search_history);
+};
+
+/*
+ * a container to RGWSyncTraceNodeRef, responsible to keep track
+ * of live nodes, and when last ref is dropped, calls ->finish()
+ * so that node moves to the retired list in the manager
+ */
+class RGWSyncTraceNodeContainer {
```

Comment:
this wrapper adds an extra allocation to create and indirection to access. consider using a custom deleter with `shared_ptr` instead? this works for me:
```patch
-using RGWSTNCRef = std::shared_ptr<RGWSyncTraceNodeContainer>;
 
-RGWSTNCRef RGWSyncTraceManager::add_node(RGWSyncTraceNode *node)
+RGWSyncTraceNodeRef RGWSyncTraceManager::add_node(RGWSyncTraceNode *node)
 {
   RWLock::WLocker wl(lock);
   RGWSyncTraceNodeRef& ref = nodes[node->handle];
   ref.reset(node);
-  return RGWSTNCRef(new RGWSyncTraceNodeContainer(ref));
+  // return a separate shared_ptr that calls finish() on the node instead of
+  // deleting it. the lambda capture holds a reference to the original 'ref'
+  auto deleter = [ref] (RGWSyncTraceNode *node) { node->finish(); };
+  return {node, deleter};
 }
```